### PR TITLE
#200: Datasource packetfabric_locations_cloud

### LINKED ISSUE #200: data source packetfabric_locations_cloud
List all PacketFabric cloud locations

data source ``packetfabric_locations_cloud``
GET [/v2/locations/cloud](https://docs.packetfabric.com/api/v2/swagger/#/Locations/get_cloud_location_list)

Checklist:
- [ ] tested locally
- [ ] added/updated acceptance tests
- [ ] add examples (in example folder)
- [ ] generate docs

### PR DESCRIPTION
**Pull request by pontovinte**
Signed-off-by: Yrineu Rodrigues <yrineu@pontovinte.com>

## Description

Closes #200 

## Checklist

- [x] tested locally
- [x] added/updated acceptance tests
- [x] add examples (in example folder)
- [x] generate docs

### CODE REVIEW

**Comment by medzin** [comment]
Acceptance test added.

File: docs/guides/importing.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ resource "packetfabric_cloud_router" "awesome_cloud_routers" {
   provider     = packetfabric
   asn          = 4556
   name         = "Awesome Cloud Routers"
+  account_uuid = var.pf_account_uuid
```
Comment (elXaptation) [inline docs/guides/importing.md:25]: @pontovinte @rjouhann Not sure why doc-gen adds this.. If I recall this wasn't required. Please confirm

File: docs/resources/packetfabric_backbone_virtual_circuit.md [hunk 2]
```diff
@@ -52,6 +52,7 @@ resource "packetfabric_backbone_virtual_circuit" "vc1" {
     vlan            = var.pf_vc_vlan2
   }
   bandwidth {
+    account_uuid      = var.pf_account_uuid
```
Comment (elXaptation) [inline docs/resources/packetfabric_backbone_virtual_circuit.md:55]: same

File: docs/resources/packetfabric_backbone_virtual_circuit_marketplace.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ resource "packetfabric_backbone_virtual_circuit_marketplace" "vc_marketplace_con
     vlan            = var.pf_vc_vlan1
   }
   bandwidth {
+    account_uuid      = var.pf_account_uuid
```
Comment (elXaptation) [inline docs/resources/packetfabric_backbone_virtual_circuit_marketplace.md:41]: same

File: docs/resources/packetfabric_ix_virtual_circuit_marketplace.md [hunk 4]
```diff
@@ -38,6 +38,7 @@ resource "packetfabric_ix_virtual_circuit_marketplace" "ix_marketplace_conn1" {
     vlan            = var.pf_vc_vlan1
   }
   bandwidth {
+    account_uuid      = var.pf_account_uuid
```
Comment (elXaptation) [inline docs/resources/packetfabric_ix_virtual_circuit_marketplace.md:41]: same

**Review by elXaptation — APPROVED** [review]
One question about Account-id..

File: docs/guides/importing.md (hunk 1 above)
Comment (rjouhann) [reply docs/guides/importing.md:25]: `account_uuid` is a required parameter, however, I added the option to set it up with an env variable in v0.4.2 so we could actually omit it from the example (https://github.com/PacketFabric/terraform-provider-packetfabric/pull/272)

File: docs/resources/packetfabric_backbone_virtual_circuit.md (hunk 2 above)
Comment (rjouhann) [reply docs/resources/packetfabric_backbone_virtual_circuit.md:55]: `account_uuid` is a required parameter, however, I added the option to set it up with an env variable in `v0.4.2` so we could actually remove it from the example (https://github.com/PacketFabric/terraform-provider-packetfabric/pull/272). This is how we recommend the user to set `PF_ACCOUNT_ID` in the [getting started page](https://registry.terraform.io/providers/PacketFabric/packetfabric/latest/docs/guides/getting_started) now.

File: docs/resources/packetfabric_backbone_virtual_circuit_marketplace.md (hunk 3 above)
Comment (rjouhann) [reply docs/resources/packetfabric_backbone_virtual_circuit_marketplace.md:41]: `account_uuid` is a required parameter, however, I added the option to set it up with an env variable in `v0.4.2` so we could actually remove it from the example (https://github.com/PacketFabric/terraform-provider-packetfabric/pull/272). This is how we recommend the user to set `PF_ACCOUNT_ID` in the [getting started page](https://registry.terraform.io/providers/PacketFabric/packetfabric/latest/docs/guides/getting_started) now.

File: docs/resources/packetfabric_ix_virtual_circuit_marketplace.md (hunk 4 above)
Comment (rjouhann) [reply docs/resources/packetfabric_ix_virtual_circuit_marketplace.md:41]: `account_uuid` is a required parameter, however, I added the option to set it up with an env variable in `v0.4.2` so we could actually remove it from the example (https://github.com/PacketFabric/terraform-provider-packetfabric/pull/272). This is how we recommend the user to set `PF_ACCOUNT_ID` in the [getting started page](https://registry.terraform.io/providers/PacketFabric/packetfabric/latest/docs/guides/getting_started) now.

File: docs/resources/packetfabric_point_to_point.md [hunk 5]
```diff
@@ -18,6 +18,7 @@ resource "packetfabric_point_to_point" "ptp1" {
   description       = var.pf_description
   speed             = var.pf_ptp_speed
   media             = var.pf_ptp_media
+  account_uuid      = var.pf_account_uuid
```
[the remark this answers is not in the document]
Comment (rjouhann) [reply docs/resources/packetfabric_point_to_point.md:21]: `account_uuid` is a required parameter, however, I added the option to set it up with an env variable in `v0.4.2` so we could actually remove it from the example (https://github.com/PacketFabric/terraform-provider-packetfabric/pull/272). This is how we recommend the user to set `PF_ACCOUNT_ID` in the [getting started page](https://registry.terraform.io/providers/PacketFabric/packetfabric/latest/docs/guides/getting_started) now.

File: docs/guides/importing.md (hunk 1 above)
Comment (rjouhann) [reply docs/guides/importing.md:25]: @elXaptation I would just remove `account_uuid = var.pf_account_uuid` from the examples.

**Comment by elXaptation** [comment]
So.. sounds like I should remove it from the generated docs.. Yes?

**Comment by rjouhann** [comment]
> So.. sounds like I should remove it from the generated docs.. Yes?

yes, we should remove it from the examples under the `examples/resources` folder and from the docs. Thanks!

**Comment by elXaptation** [comment]
Okay.. Done..

**Comment by rjouhann** [comment]
> Okay.. Done..

@elXaptation I see you removed `account_uuid` in this PR https://github.com/PacketFabric/terraform-provider-packetfabric/pull/274, don't you need also to remove it from this current PR #270?

File: docs/data-sources/packetfabric_locations_cloud.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  value = data.packetfabric_locations_cloud.cloud_location_1  
+}
+```
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `cloud_connection_type` (String) Filter locations by cloud connection type.
```
Comment (rjouhann) [inline docs/data-sources/packetfabric_locations_cloud.md:29]: @elXaptation Could we add the available options: ``aws, azure, packet, google, ibm, oracle, salesforce, webex`` as per [API spec](https://docs.packetfabric.com/api/v2/swagger/#/Locations/get_cloud_location_list).

File: docs/data-sources/packetfabric_locations_cloud.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+- `cloud_provider` (String) Filter locations by cloud provider.
```
Comment (rjouhann) [inline docs/data-sources/packetfabric_locations_cloud.md:30]: @elXaptation Could we add the available options: ``hosted, dedicated`` as per [API spec](https://docs.packetfabric.com/api/v2/swagger/#/Locations/get_cloud_location_list).

**Comment by rjouhann** [comment]
@pontovinte could you also add the other options which are essential for this data-source (even though optional in the [API spec](https://docs.packetfabric.com/api/v2/swagger/#/Locations/get_cloud_location_list)):

`nat_capable`, `has_cloud_router`, `any_type`, `region`, `market`, `state`, `city`, `pop`.

One of the use case is to use this data-source to get the available zone for a pop and use one of the zone listed to create a Cloud Router Connection or a Hosted Cloud.

Another use case would be to find the pop name located in a specific city and use it to create the cloud router connection of hosted cloud.

**Review by rjouhann — CHANGES_REQUESTED** [review]

File: internal/provider/datasource_cloud_locations.go [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+				Description:  "Filter locations by cloud provider.",
+			},
+			"cloud_connection_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{"hosted", "dedicated"}, true),
+				Description:  "Filter locations by cloud connection type.",
+			},
+			"cloud_locations": {
+				Type:     schema.TypeList,
```
Comment (rjouhann) [inline internal/provider/datasource_cloud_locations.go:29]: @pontovinte @elXaptation could we add a description for cloud_locations?

Something like `Description: "The list of list of physical locations optionally filtered by provided parameters.",`

File: internal/provider/provider.go [hunk 9]
```diff
@@ -132,6 +132,7 @@ func Provider() *schema.Provider {
 			"packetfabric_marketplace_service_requests":  dataSourceVcRequests(),
 			"packetfabric_virtual_circuits":              datasourceBackboneServices(),
 			"packetfabric_point_to_point":                datasourcePointToPoint(),
+			"packetfabric_locations_cloud":               datasourceCloudLoctions(),
```
Comment (rjouhann) [inline internal/provider/provider.go:135]: `datasourceCloudLoctions` should be `dataSourceCloudLoctions`

**Review by rjouhann — CHANGES_REQUESTED** [review]

**Review by rjouhann — DISMISSED** [review]

**Review by caitlinwheeless — DISMISSED** [review]
No other doc changes needed

**Comment by rjouhann** [comment]
@caitlinwheeless @elXaptation  could you regenerate the docs?

The new filter @pontovinte added aren't showing.

Also, could we add the options for `cloud_connection_type` (aws, azure, packet, google, ibm, oracle, salesforce, webex) and `cloud_provider` (dedicated or hosted)?

**Comment by caitlinwheeless** [comment]
> @caitlinwheeless @elXaptation could you regenerate the docs?
> 
> The new filter @pontovinte added aren't showing.
> 
> Also, could we add the options for `cloud_connection_type` (aws, azure, packet, google, ibm, oracle, salesforce, webex) and `cloud_provider` (dedicated or hosted)?

I can regenerate the docs, but it might not make sense to do that until after you've added all the options.

**Review by rjouhann — APPROVED** [review]

**Review by rjouhann — APPROVED** [review]

**Pull request merged**